app/controllers/duties_controller.rb

```diff
@@ -1,2 +1,45 @@
 class DutiesController < ApplicationController
+  def index; end
+
+  def process_grab
+    @user = User.find(params[:user_id])
+    @duty = Duty.find(params[:duty_id])
+
+    # user_id of duty should and must be nil
+    if @duty.user.nil? && @duty.update(user: @user)
+      flash[:notice] = 'Successfully grabbed duty!'
+    else
+      flash[:alert] = 'Failed to grab duty!'
+    end
+    redirect_to duties_path
+  end
+
+  def process_drop
+    @duty = Duty.find(params[:duty_id])
+
+    if @duty.update(user: nil)
+      flash[:notice] = 'Successfully dropped duty!'
+    else
+      flash[:alert] = 'Failed to drop duty!'
+    end
+    redirect_to duties_path
+  end
+
+  def edit # mass edit
+    @duties = Duty.all
+    @timeslots = Timeslot.all.order(:start_time)
+    @places = Place.all
+    week_offset = params[:week_offset]
+    day_offset = (Time.zone.now.wday - 1) % 7
+    start_date = Time.zone.now - day_offset.days + week_offset.weeks
+    @duties = Duty.where("date >= #{start_date} && " \
```

Comment:
I think this should be abstracted out to be a model function, something like `Duty.week_of(Time.zone.now - 7.days)`?